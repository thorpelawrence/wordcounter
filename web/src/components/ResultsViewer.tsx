import { useMemo } from "react";
import { Column, useTable, useSortBy } from "react-table";

export interface WordCounts {
  word: string;
  count: number;
}

const ResultsViewer = ({ data }: { data: WordCounts[] }) => {
  const columns = useMemo<Column<WordCounts>[]>(
    () => [
      { Header: "Word", accessor: "word" },
      { Header: "Occurrences", accessor: "count" },
    ],
    []
  );

  const sortBy = useMemo(() => [{ id: "count", desc: true }], []);

  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
    useTable(
      {
        data,
        columns,
        initialState: {
          // @ts-ignore
          sortBy,
        },
      },
      useSortBy
    );

  const rowsSlice = rows.slice(0, 100) // TODO optimise, pagination etc instead

  return (
    <table {...getTableProps()}>
      <thead>
        {headerGroups.map((headerGroup) => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map((column: any) => (
              <th {...column.getHeaderProps(column.getSortByToggleProps())}>
                {column.render("Header")}
                <span>
                  {column.isSorted ? (column.isSortedDesc ? " ðŸ”½" : " ðŸ”¼") : ""}
                </span>
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rowsSlice.map((row, i) => {
          prepareRow(row);
          return (
            <tr {...row.getRowProps()}>
              {row.cells.map((cell) => {
                return <td {...cell.getCellProps()}>{cell.render("Cell")}</td>;
              })}
            </tr>
          );
        })}
      </tbody>
    </table>
  );
};

export default ResultsViewer;
