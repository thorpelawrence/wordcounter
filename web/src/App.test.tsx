import { render, screen } from "@testing-library/react";
import ResultsViewer from "./components/ResultsViewer";

test("renders correct data", async () => {
  render(
    <ResultsViewer
      data={{
        hello: 1,
        world: 2,
        test: 3,
      }}
    />
  );
  expect(screen.getByText(/hello/)).toBeInTheDocument();
  expect(screen.getByText(/test/)).toBeInTheDocument();
  expect(screen.getByRole("table")).toMatchSnapshot();
});
