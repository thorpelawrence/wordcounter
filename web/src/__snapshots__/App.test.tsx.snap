// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct data 1`] = `
<table
  role="table"
>
  <thead>
    <tr
      role="row"
    >
      <th
        colspan="1"
        role="columnheader"
        style="cursor: pointer;"
        title="Toggle SortBy"
      >
        Word
        <span />
      </th>
      <th
        colspan="1"
        role="columnheader"
        style="cursor: pointer;"
        title="Toggle SortBy"
      >
        Occurrences
        <span>
           ðŸ”½
        </span>
      </th>
    </tr>
  </thead>
  <tbody
    role="rowgroup"
  >
    <tr
      role="row"
    >
      <td
        role="cell"
      >
        test
      </td>
      <td
        role="cell"
      >
        3
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        role="cell"
      >
        world
      </td>
      <td
        role="cell"
      >
        2
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        role="cell"
      >
        hello
      </td>
      <td
        role="cell"
      >
        1
      </td>
    </tr>
  </tbody>
</table>
`;
